### ISSUE TITLE
PEP 582 support

### ISSUE THREAD

**Issue by danieleades** [python-poetry/poetry#3691]
basically i'm plugging a stale issue - #872

[PEP 582](https://www.python.org/dev/peps/pep-0582/) is continuing to gain traction, and looks like a really nice solution for having an isolated environment.

![xkcd](https://imgs.xkcd.com/comics/standards.png)

A couple of new package managers have sprung up around this standard. I think that Poetry should abstract over the package management strategy, and support *either* a virtual environment or a PEP582-compliant `_pypackages__/` directory.

first-class IDE support isn't there yet, but it doesn't take a lot of configuration to get it working for vs code.

## [PDM](https://pdm.fming.dev/)

PDM is a python-based package manager which bills itself as an evolution of Poetry. It looks really mature, the maintainer is very responsive. I've been playing around with it and haven't managed to break it yet. The feature set and CLI is *very* similar to what Poetry offers.

PDM offers the ability to switch python interpreters on the fly.

I've not found any issues with the dependency solving, but it would be nice to know which library has the edge there.

## [PyFlow](https://github.com/David-OConnor/pyflow)

In general it seems a little more minimalist than Poetry or PDM in terms of its API surface, and perhaps less mature than PDM. In my local testing i ran into quite a few bugs, so I don't think it's quite there yet. Seems to be actively developed though.

PyFlow is written in Rust- I'm a massive Rust fan-boy, so that appeals to me. I would think that if this project matures, dependency solving is likely to be *dramatically* faster in Rust than python.

PyFlow also allows for easily switching python interpreters on the fly.

using `__pypackages__/` affords a couple of neat features you don't get with a virtual environment- these can be *nested* recursively, which allows a dependency to have it's own isolated dependencies. PyFlow claims to take advantage of this to be able to install multiple versions of the same dependency as a fallback 'trick' for dependency resolution. I've not looked at this in detail, but I'm guessing this would fall apart if any breaking changes from those isolated transitive dependency's were re-exported or exposed in the public API of the dependency... Nevertheless, it's a clever solution!

---

So far the only 'killer feature' I see is the ability to easily switch python versions. That's a big help for testing, as well as development. Though I can certainly see how this can also simplify Docker deployments or dependency vendoring too.

**Comment by Yajo** [comment]
PEP 582 approach has some benefits. For me, there are a few tools I love to use for development, such as https://github.com/gruns/icecream or https://github.com/prompt-toolkit/ptpython, but I don't like to fill every project's dev dependencies with them. So I don't really have them available, or I have to use pip to install them.

Well, by not using a virtualenv, that makes the app have the libs it requires, while letting the dev configure his system as he wants. That's really nice.

**Comment by sdispater** [comment]
PEP 582 is still in draft and there is still no consensus on whether it is a good idea and if it will ever be supported at the interpreter level.

PEP 582 is currently a bet at best and that's a bet I am not willing to take for the project and its users.

That being said once the plugin system is implemented for Poetry (which will be available in the 1.2 release) nothing will prevent community support for it.

> PEP 582 is continuing to gain traction

Apart from the two projects you mentioned here I don't think it is really used in the wild and its usage is marginal at best. And its support in the tooling and ecosystem is close to non-existent and rightly so since it's not an accepted standard.

> PDM is a python-based package manager which bills itself as an evolution of Poetry.

Except for PEP-582, it's almost a perfect copy of Poetry (even the lock file is extremely similar) so I am interested to see where the added value is.

> PDM offers the ability to switch python interpreters on the fly.
> PyFlow also allows for easily switching python interpreters on the fly.

I am not sure what that means, honestly.

So to summarize, there is no plan for Poetry to add support for PEP 582 unless it becomes an accepted standard.

For the time being, the project will focus on virtual environments which are pretty much standard and are wildly supported.

As a side-note:

> I would think that if this project matures, dependency solving is likely to be dramatically faster in Rust than python.

It won't – at least not in a significant magintude – because currently the bottleneck is not on the algorithm or even the language but on the required download and package introspection needed to retrieve the packages metadata.

**Comment by danieleades** [comment]
@sdispater thanks for the summary

I guess we'll continue to watch this space to see if PEP 582 sees wider adoption, and if these new package managers can leverage it to provide differentiating features.

**Comment by sdispater** [comment]
@danieleades Just to be clear: I also think PEP 582 could really be an improvement however since it's not really standardized yet it might introduce more complexity than necessary. That's why I think our time, as maintainers, will be better spent elsewhere for now (there is a lot of new features planned already for the next version).

But if PEP 582 is ever accepted I will be the first to add support for it in Poetry.

**Comment by danieleades** [comment]
I've just run into a real-life use-case for PEP 582

i use Poetry to minimise the number of system dependencies that I need to play with. That makes testing consistent, doesn't mess with my system packages, and makes on-boarding easier for other developers working on the same codebase.

I'm finding that the number of system-wide dependencies i'm using is starting to increase- I'm now using Poetry, Nox, and Nox-Poetry on a particular project.

Now it might be possible to install those using Poetry, and bootstrap the test dependencies that way, but this causes me dependency issues. The extra packages make the dependency resolution overly-constrained (in my case Nox relies on some recent packages that break some of my less up-to-date dependencies).

If PEP 582 were in play, Nox could have its own dependencies installed recursively in a nested `__pypackages__` directory, without constraining my other dependencies. Using a virtualenv forces all of your dependencies to play nicely together in one environment, which is far more restrictive.

**Comment by Alphasite** [comment]
This seens to be a rough benchmark from what I can tell, but if nothing else PDM seems to have a very fast resolver: https://dev.to/frostming/a-review-pipenv-vs-poetry-vs-pdm-39b4

**Comment by danieleades** [comment]
> This seens to be a rough benchmark from what I can tell, but if nothing else PDM seems to have a very fast resolver: https://dev.to/frostming/a-review-pipenv-vs-poetry-vs-pdm-39b4

i've done a bit of hacking in the `PDM` codebase and I get a very strong impression that it started life as a copy and paste of `Poetry` (some of the modules at least).

I hate seeing this kind of 'competitiveness' between projects that have massive similarities. I would *love* to see elements of core logic extracted into libraries and shared between the two.

**Comment by Alphasite** [comment]
From the UX perspective I wouldn’t be too surprised, but to clarify I meant share this more as a maybe there are some nice optimisations and learnings which can be shared from the PDM codebase.

**Comment by bolinocroustibat** [comment]
I strongly hope PEP 582 will become standard. I can't wait for Python standards to find a better solution to virtual environments. We are all very used to them and see them as an improvement over the no-venv workflow, but we forgot how it started as a band-aid, and how most of other modern languages dependency management flow is immensly more normal and convenient (Cargo/Rust, NPM/JS, etc.).

**Comment by RitikShah** [comment]
Now that Poetry 1.2 has officially released 🎉, has there been any work on a plugin for this feature. I think a great way to validate the pep would be to utilize it in some projects and I'm too fond of `poetry` to leave it for `pyflow` or `pdm`.

Also plugins are a cool way to validate new potential features!

**Comment by neersighted** [comment]
`poetry-bundle-plugin` is the best place to introduce this functionality, I would think

**Comment by neersighted** [comment]
I'm closing this out of scope for Poetry itself -- given that PEP 582 is stalled, I don't think it's useful to reason about too much right now. I am incorrect on bundle being the right place for this to live after a re-read -- native support would likely be useful here. However, we'd have to see this feature in a released Python version to even start considering it for inclusion.
